File: examples/university-demo/src/main/java/io/axoniq/demo/university/faculty/write/CourseId.java
package io.axoniq.demo.university.faculty.write;

import java.util.UUID;

public record CourseId(String raw) {

    private final static String ENTITY_TYPE = "Course";

    public CourseId {
        if (raw == null || raw.isBlank()) {
            throw new IllegalArgumentException("Course ID cannot be null or empty");
        }
        raw = withType(raw);
    }

    public static CourseId of(String raw) {
        return new CourseId(raw);
    }

    public static CourseId random() {
        return new CourseId(UUID.randomUUID().toString());
    }

    @Override
    public String toString() {
        return raw;
    }

    private static String withType(String id) {
        return id.startsWith(ENTITY_TYPE + ":") ? id : ENTITY_TYPE + ":" + id;
    }
}

File: examples/university-demo-kotlin/src/main/kotlin/shared/ids/CourseId.kt
@file:OptIn(ExperimentalStdlibApi::class)

package org.axonframework.examples.university.shared.ids

import java.util.*

@JvmExposeBoxed
@JvmInline
value class CourseId(val value: String) {
    companion object {
        fun random(): CourseId = CourseId(UUID.randomUUID().toString())
    }

    override fun toString() = value
}
